File: src/OCaml/lib/finger_tree/priority_search_queue.ml
(** Implementation of the persistent key-ordered priority-search queue with a cached priority-then-
    key winner.

    The representation is the winner-cached key-ordered AVL every sibling workspace ships: a
    path-copied AVL tree ordered by key in which every node also caches the minimum-priority entry
    of its whole subtree (ties broken by key order). The key index answers keyed reads and writes in
    O(log n); the winner caches answer the minimum-priority question at the root in O(1) with no
    search; and a keyed delete of the cached winner gives an O(log n) delete-min. Each node also
    caches its subtree size, so rank selection and key-rank bounds are O(log n) size-directed
    descents. *)

type ('key, 'priority, 'value) entry = { key : 'key; priority : 'priority; value : 'value }

type ('key, 'priority, 'value) node = {
  entry : ('key, 'priority, 'value) entry;
  left : ('key, 'priority, 'value) node option;
  right : ('key, 'priority, 'value) node option;
  winner : ('key, 'priority, 'value) entry;
  height : int;
  size : int;
}

type ('key, 'priority, 'value) t = {
  key_order : 'key Common.Comparator.t;
  priority_order : 'priority Common.Comparator.t;
  root : ('key, 'priority, 'value) node option;
  recomputation_count : int;
}

type statistics = { count : int; estimated_height : int; winner_recomputations : int }

let empty ~key_comparator ~priority_comparator =
  {
    key_order = key_comparator;
    priority_order = priority_comparator;
    root = None;
    recomputation_count = 0;
  }

let node_size = function None -> 0 | Some node -> node.size
let node_height = function None -> 0 | Some node -> node.height
let count queue = node_size queue.root
let is_empty queue = count queue = 0
let entry_key entry = entry.key
let entry_priority entry = entry.priority
let entry_value entry = entry.value
let compare_key queue = Common.Comparator.compare queue.key_order

let before queue left right =
  let by_priority = Common.Comparator.compare queue.priority_order left.priority right.priority in
  by_priority < 0 || (by_priority = 0 && compare_key queue left.key right.key < 0)

(* The subtree winner: the node's own entry unless a child subtree's cached winner beats it. One
   cached winner per node is what makes the root's winner the whole queue's minimum without a
   search. *)
let select_winner queue entry left right =
  let winner = match left with Some l when before queue l.winner entry -> l.winner | _ -> entry in
  match right with Some r when before queue r.winner winner -> r.winner | _ -> winner

(* Every path-copied node recomputes its winner cache; [recomputations] counts them so the
   statistics report how much winner maintenance a lineage has performed. *)
let make queue recomputations entry left right =
  incr recomputations;
  {
    entry;
    left;
    right;
    winner = select_winner queue entry left right;
    height = 1 + Int.max (node_height left) (node_height right);
    size = 1 + node_size left + node_size right;
  }

let rotate_left queue recomputations node =
  match node.right with
  | None -> invalid_arg "priority-search left rotation requires a right child"
  | Some pivot ->
      make queue recomputations pivot.entry
        (Some (make queue recomputations node.entry node.left pivot.left))
        pivot.right

let rotate_right queue recomputations node =
  match node.left with
  | None -> invalid_arg "priority-search right rotation requires a left child"
  | Some pivot ->
      make queue recomputations pivot.entry pivot.left
        (Some (make queue recomputations node.entry pivot.right node.right))

let balance queue recomputations node =
  let factor = node_height node.left - node_height node.right in
  if factor > 1 then
    match node.left with
    | None -> invalid_arg "priority-search left imbalance lacks a left child"
    | Some left ->
        if node_height left.left < node_height left.right then
          rotate_right queue recomputations
            (make queue recomputations node.entry
               (Some (rotate_left queue recomputations left))
               node.right)
        else rotate_right queue recomputations node
  else if factor < -1 then
    match node.right with
    | None -> invalid_arg "priority-search right imbalance lacks a right child"
    | Some right ->
        if node_height right.right < node_height right.left then
          rotate_left queue recomputations
            (make queue recomputations node.entry node.left
               (Some (rotate_right queue recomputations right)))
        else rotate_left queue recomputations node
  else node

let publish queue recomputations root =
  { queue with root; recomputation_count = queue.recomputation_count + !recomputations }

let rec find_node queue key = function
  | None -> None
  | Some node ->
      let ordering = compare_key queue key node.entry.key in
      if ordering = 0 then Some node.entry
      else if ordering < 0 then find_node queue key node.left
      else find_node queue key node.right

let find key queue = find_node queue key queue.root
let mem key queue = Option.is_some (find key queue)

let nth index queue =
  if index < 0 || index >= count queue then None
  else
    let rec select index = function
      | None -> None
      | Some node ->
          let left_size = node_size node.left in
          if index < left_size then select index node.left
          else if index = left_size then Some node.entry
          else select (index - left_size - 1) node.right
    in
    select index queue.root

(* The rank of the boundary before ([upper] = false) or after ([upper] = true) the probe's
   equivalence class: an O(log n) size-directed descent. *)
let bound_rank upper key queue =
  let rec descend rank = function
    | None -> rank
    | Some node ->
        let ordering = compare_key queue node.entry.key key in
        if ordering < 0 || (upper && ordering = 0) then
          descend (rank + node_size node.left + 1) node.right
        else descend rank node.left
  in
  descend 0 queue.root

let lower_bound key queue = bound_rank false key queue
let upper_bound key queue = bound_rank true key queue

let rec insert_node queue recomputations entry ~overwrite = function
  | None -> make queue recomputations entry None None
  | Some node ->
      let ordering = compare_key queue entry.key node.entry.key in
      if ordering = 0 then
        if overwrite then
          (* Replacement keeps the stored key representative and re-seats the entry in the winner
             caches along this path. *)
          make queue recomputations
            { key = node.entry.key; priority = entry.priority; value = entry.value }
            node.left node.right
        else invalid_arg "priority-search insertion reached an existing key"
      else if ordering < 0 then
        balance queue recomputations
          (make queue recomputations node.entry
             (Some (insert_node queue recomputations entry ~overwrite node.left))
             node.right)
      else
        balance queue recomputations
          (make queue recomputations node.entry node.left
             (Some (insert_node queue recomputations entry ~overwrite node.right)))

let add key priority value queue =
  if mem key queue then Error "priority-search queue already contains an equivalent key"
  else
    let recomputations = ref 0 in
    let root = insert_node queue recomputations { key; priority; value } ~overwrite:false queue.root in
    Ok (publish queue recomputations (Some root))

let set key priority value queue =
  let added = not (mem key queue) in
  let recomputations = ref 0 in
  let root = insert_node queue recomputations { key; priority; value } ~overwrite:true queue.root in
  (added, publish queue recomputations (Some root))

let rec extract_minimum queue recomputations node =
  match node.left with
  | None -> (node.entry, node.right)
  | Some left ->
      let minimum, rest = extract_minimum queue recomputations left in
      (minimum, Some (balance queue recomputations (make queue recomputations node.entry rest node.right)))

let rec remove_node queue recomputations key = function
  | None -> None
  | Some node -> (
      let ordering = compare_key queue key node.entry.key in
      if ordering = 0 then
        match (node.left, node.right) with
        | None, other | other, None -> Some (node.entry, other)
        | Some _, Some right ->
            let successor, rest = extract_minimum queue recomputations right in
            Some
              ( node.entry,
                Some (balance queue recomputations (make queue recomputations successor node.left rest))
              )
      else if ordering < 0 then
        Option.map
          (fun (removed, left) ->
            (removed, Some (balance queue recomputations (make queue recomputations node.entry left node.right))))
          (remove_node queue recomputations key node.left)
      else
        Option.map
          (fun (removed, right) ->
            (removed, Some (balance queue recomputations (make queue recomputations node.entry node.left right))))
          (remove_node queue recomputations key node.right))

let remove key queue =
  let recomputations = ref 0 in
  Option.map
    (fun (removed, root) -> (removed, publish queue recomputations root))
    (remove_node queue recomputations key queue.root)

let minimum queue = Option.map (fun node -> node.winner) queue.root

let minimum_view queue =
  match minimum queue with
  | None -> None
  | Some entry -> Option.map (fun (_, successor) -> (entry, successor)) (remove entry.key queue)

(* Query frames for the bounded enumeration: a node whose subtree may still qualify, or an entry
   already known to qualify and waiting its turn in key order. The explicit stack bounds traversal
   space by the tree height, mirroring the reference implementation frame for frame. *)
type ('key, 'priority, 'value) query_frame =
  | Visit of ('key, 'priority, 'value) node
  | Emit of ('key, 'priority, 'value) entry

let enumerate_at_most ~minimum_key ~maximum_key ~maximum_priority queue =
  if compare_key queue minimum_key maximum_key > 0 then
    Error "priority-search queue range has its minimum key after its maximum key"
  else begin
    let compare_priority = Common.Comparator.compare queue.priority_order in
    let rec drain pending emitted =
      match pending with
      | [] -> List.rev emitted
      | Emit entry :: rest -> drain rest (entry :: emitted)
      | Visit node :: rest ->
          (* The winner cache is the pruning structure: when even this subtree's best priority
             exceeds the threshold, nothing below can qualify and the whole subtree is skipped. *)
          if compare_priority node.winner.priority maximum_priority > 0 then drain rest emitted
          else begin
            let above_minimum = compare_key queue node.entry.key minimum_key in
            let below_maximum = compare_key queue node.entry.key maximum_key in
            let pending = rest in
            let pending =
              match node.right with
              | Some right when below_maximum < 0 -> Visit right :: pending
              | _ -> pending
            in
            let pending =
              if
                above_minimum >= 0 && below_maximum <= 0
                && compare_priority node.entry.priority maximum_priority <= 0
              then Emit node.entry :: pending
              else pending
            in
            let pending =
              match node.left with
              | Some left when above_minimum > 0 -> Visit left :: pending
              | _ -> pending
            in
            drain pending emitted
          end
    in
    match queue.root with
    | None -> Ok []
    | Some root -> Ok (drain [ Visit root ] [])
  end

let entries_by_key queue =
  let rec walk node acc =
    match node with
    | None -> acc
    | Some node -> walk node.left (node.entry :: walk node.right acc)
  in
  walk queue.root []

let statistics queue =
  {
    count = count queue;
    estimated_height = node_height queue.root;
    winner_recomputations = queue.recomputation_count;
  }

File: src/Kotlin/FingerTree/src/tools/datastructures/fingertree/PrioritySearchQueue.kt
package tools.datastructures.fingertree

import java.util.IdentityHashMap
import kotlin.math.abs
import kotlin.math.max

private fun <T> psqValuesEqual(left: T, right: T): Boolean = left === right || left == right

/** One unique ordered key, its priority, and its payload. Lower priorities win. */
public data class PrioritySearchEntry<K, P, V>(
    val key: K,
    val priority: P,
    val value: V,
)

/** Statistics returned by the independent winner-cached AVL audit. */
public data class PrioritySearchQueueStatistics(
    val count: Int,
    val height: Int,
    val maximumAbsoluteBalanceFactor: Int,
)

/** Result of a non-overwriting keyed insertion. */
public data class PrioritySearchAddResult<K, P, V>(
    val added: Boolean,
    val queue: PrioritySearchQueue<K, P, V>,
)

/** Result of a keyed removal. [entry] is non-null exactly when [removed] is true. */
public data class PrioritySearchRemoveResult<K, P, V>(
    val removed: Boolean,
    val entry: PrioritySearchEntry<K, P, V>?,
    val queue: PrioritySearchQueue<K, P, V>,
)

/** One successful minimum deletion. */
public data class PrioritySearchMinimumView<K, P, V>(
    val entry: PrioritySearchEntry<K, P, V>,
    val remainder: PrioritySearchQueue<K, P, V>,
)

/**
 * A persistent ordered map whose AVL nodes cache their minimum-priority entry.
 *
 * One entry is retained per [keyComparator] equivalence class. Equal priorities break by retained
 * key order, making [minimum] deterministic. Keyed updates and minimum deletion are O(log n),
 * minimum lookup is O(1), and [enumerateAtMost] prunes a subtree when its cached winner already
 * exceeds the threshold.
 */
public class PrioritySearchQueue<K, P, V> private constructor(
    private val root: Node<K, P, V>?,
    public val keyComparator: Comparator<in K>,
    public val priorityComparator: Comparator<in P>,
) : Iterable<PrioritySearchEntry<K, P, V>> {
    public companion object {
        /** Creates an empty queue with shared natural-order comparators. */
        public fun <K : Comparable<K>, P : Comparable<P>, V> empty(): PrioritySearchQueue<K, P, V> =
            PrioritySearchQueue(root = null, naturalOrder(), naturalOrder())

        /** Creates an empty queue retaining the supplied comparison policies. */
        public fun <K, P, V> empty(
            keyComparator: Comparator<in K>,
            priorityComparator: Comparator<in P>,
        ): PrioritySearchQueue<K, P, V> =
            PrioritySearchQueue(root = null, keyComparator, priorityComparator)

        /** O(n log n) last-wins construction using natural ordering. */
        public fun <K : Comparable<K>, P : Comparable<P>, V> from(
            entries: Iterable<PrioritySearchEntry<K, P, V>>,
        ): PrioritySearchQueue<K, P, V> = from(entries, naturalOrder(), naturalOrder())

        /** O(n log n) last-wins construction retaining explicit comparators. */
        public fun <K, P, V> from(
            entries: Iterable<PrioritySearchEntry<K, P, V>>,
            keyComparator: Comparator<in K>,
            priorityComparator: Comparator<in P>,
        ): PrioritySearchQueue<K, P, V> {
            var result = empty<K, P, V>(keyComparator, priorityComparator)
            for (entry in entries) {
                result = result.setItem(entry.key, entry.priority, entry.value)
            }
            return result
        }
    }

    public val count: Int
        get() = root?.count ?: 0

    public val size: Int
        get() = count

    public val isEmpty: Boolean
        get() = root == null

    /** Cached AVL height for diagnostics and logarithmic-bound audits. */
    public val height: Int
        get() = root?.height ?: 0

    /** The global minimum-priority entry in O(1). */
    public val minimum: PrioritySearchEntry<K, P, V>
        get() = root?.winner ?: throw NoSuchElementException("The priority search queue is empty.")

    public fun minimumOrNull(): PrioritySearchEntry<K, P, V>? = root?.winner

    public fun containsKey(key: K): Boolean = findNode(key) != null

    /** Returns the stored entry and therefore the first retained key representative. */
    public fun getEntryOrNull(key: K): PrioritySearchEntry<K, P, V>? = findNode(key)?.entry

    /**
     * Adds or replaces an entry in O(log n). An equivalent key retains its first concrete key.
     * Replacement is a no-op only when priority ordering, priority equality, and payload equality
     * all agree with the stored representation.
     */
    public fun setItem(key: K, priority: P, value: V): PrioritySearchQueue<K, P, V> {
        val result = set(root, PrioritySearchEntry(key, priority, value), overwrite = true)
        return if (result.node === root) this else PrioritySearchQueue(result.node, keyComparator, priorityComparator)
    }

    /** Attempts to add a new key without replacing an equivalent existing key. */
    public fun tryAdd(key: K, priority: P, value: V): PrioritySearchAddResult<K, P, V> {
        val result = set(root, PrioritySearchEntry(key, priority, value), overwrite = false)
        val queue = if (result.added) {
            PrioritySearchQueue(result.node, keyComparator, priorityComparator)
        } else {
            this
        }
        return PrioritySearchAddResult(result.added, queue)
    }

    /** Removes an equivalent key in O(log n), preserving identity when absent. */
    public fun remove(key: K): PrioritySearchQueue<K, P, V> {
        val result = removeNode(root, key)
        return if (result.entry == null) this else PrioritySearchQueue(result.node, keyComparator, priorityComparator)
    }

    /** Attempts keyed removal while returning the stored representative. */
    public fun tryRemove(key: K): PrioritySearchRemoveResult<K, P, V> {
        val result = removeNode(root, key)
        val removed = result.entry != null
        return PrioritySearchRemoveResult(
            removed,
            result.entry,
            if (removed) PrioritySearchQueue(result.node, keyComparator, priorityComparator) else this,
        )
    }

    /** Removes and returns the global winner in O(log n). */
    public fun deleteMinimum(): PrioritySearchMinimumView<K, P, V> {
        val entry = root?.winner ?: throw NoSuchElementException("The priority search queue is empty.")
        val result = removeNode(root, entry.key)
        check(result.entry != null) { "The cached priority winner was not present by key." }
        return PrioritySearchMinimumView(entry, PrioritySearchQueue(result.node, keyComparator, priorityComparator))
    }

    /** Returns null only when the queue is empty. */
    public fun minimumView(): PrioritySearchMinimumView<K, P, V>? =
        if (root == null) null else deleteMinimum()

    /**
     * Lazily enumerates entries in the inclusive key range whose priorities are at most the
     * inclusive threshold. Results are in key order. The cost is O(log n + v), where v is the
     * number of nodes whose subtrees winner pruning cannot exclude; v may be n.
     */
    public fun enumerateAtMost(
        minimumKey: K,
        maximumKey: K,
        maximumPriority: P,
    ): Sequence<PrioritySearchEntry<K, P, V>> {
        require(keyComparator.compare(minimumKey, maximumKey) <= 0) {
            "The minimum key must not follow the maximum key."
        }
        val snapshot = root
        return sequence {
            val actualRoot = snapshot ?: return@sequence
            val pending = ArrayList<QueryFrame<K, P, V>>()
            pending.add(QueryFrame(actualRoot, emit = false))
            while (pending.isNotEmpty()) {
                val frame = pending.removeAt(pending.lastIndex)
                val node = frame.node
                if (frame.emit) {
                    yield(node.entry)
                    continue
                }
                if (priorityComparator.compare(node.winner.priority, maximumPriority) > 0) {
                    continue
                }

                val lower = keyComparator.compare(node.entry.key, minimumKey)
                val upper = keyComparator.compare(node.entry.key, maximumKey)
                if (upper < 0) {
                    node.right?.let { pending.add(QueryFrame(it, emit = false)) }
                }
                if (lower >= 0 && upper <= 0 &&
                    priorityComparator.compare(node.entry.priority, maximumPriority) <= 0
                ) {
                    pending.add(QueryFrame(node, emit = true))
                }
                if (lower > 0) {
                    node.left?.let { pending.add(QueryFrame(it, emit = false)) }
                }
            }
        }
    }

    /** In-order iteration is explicit-stack and performs no priority comparisons. */
    override fun iterator(): Iterator<PrioritySearchEntry<K, P, V>> = object : Iterator<PrioritySearchEntry<K, P, V>> {
        private val pending = ArrayList<Node<K, P, V>>()
        private var cursor = root

        init {
            pushLeftSpine()
        }

        override fun hasNext(): Boolean = pending.isNotEmpty()

        override fun next(): PrioritySearchEntry<K, P, V> {
            if (pending.isEmpty()) {
                throw NoSuchElementException()
            }
            val node = pending.removeAt(pending.lastIndex)
            cursor = node.right
            pushLeftSpine()
            return node.entry
        }

        private fun pushLeftSpine() {
            while (cursor != null) {
                val node = cursor ?: break
                pending.add(node)
                cursor = node.left
            }
        }
    }

    /**
     * Validates strict BST bounds, AVL balance, cached count/height, and every cached winner with an
     * explicit worklist. The recursive update depth is bounded by the validated AVL height.
     */
    public fun validateStructure(): PrioritySearchQueueStatistics {
        val actualRoot = root ?: return PrioritySearchQueueStatistics(0, 0, 0)
        val pending = ArrayList<ValidationFrame<K, P, V>>()
        pending.add(ValidationFrame(actualRoot, lower = null, upper = null))
        var validatedCount = 0
        var maximumBalance = 0
        while (pending.isNotEmpty()) {
            val frame = pending.removeAt(pending.lastIndex)
            val node = frame.node
            frame.lower?.let {
                check(keyComparator.compare(node.entry.key, it.value) > 0) {
                    "A priority-search node crosses its lower key bound."
                }
            }
            frame.upper?.let {
                check(keyComparator.compare(node.entry.key, it.value) < 0) {
                    "A priority-search node crosses its upper key bound."
                }
            }

            val expectedCount = Math.addExact(1, Math.addExact(node.left?.count ?: 0, node.right?.count ?: 0))
            val expectedHeight = Math.addExact(1, max(heightOf(node.left), heightOf(node.right)))
            val balance = heightOf(node.left) - heightOf(node.right)
            check(node.count == expectedCount && node.height == expectedHeight) {
                "A priority-search node has invalid cached metadata."
            }
            check(abs(balance) <= 1) { "A priority-search node violates AVL balance." }

            val expectedWinner = selectWinner(node.entry, node.left, node.right)
            check(sameStoredEntry(expectedWinner, node.winner)) {
                "A priority-search node has an invalid cached winner."
            }

            validatedCount = Math.addExact(validatedCount, 1)
            maximumBalance = max(maximumBalance, abs(balance))
            node.right?.let {
                pending.add(ValidationFrame(it, KeyBound(node.entry.key), frame.upper))
            }
            node.left?.let {
                pending.add(ValidationFrame(it, frame.lower, KeyBound(node.entry.key)))
            }
        }
        check(validatedCount == count) {
            "Priority-search root metadata disagrees with the validated tree."
        }
        return PrioritySearchQueueStatistics(validatedCount, height, maximumBalance)
    }

    /** Counts node objects retained by identity across versions. */
    internal fun sharedNodeCount(other: PrioritySearchQueue<K, P, V>): Int {
        val identities = IdentityHashMap<Node<K, P, V>, Boolean>()
        collectNodes(root, identities)
        val visited = IdentityHashMap<Node<K, P, V>, Boolean>()
        val pending = ArrayList<Node<K, P, V>>().apply { other.root?.let(::add) }
        var shared = 0
        while (pending.isNotEmpty()) {
            val node = pending.removeAt(pending.lastIndex)
            if (visited.put(node, true) != null) {
                continue
            }
            if (identities.containsKey(node)) {
                shared++
            }
            node.left?.let(pending::add)
            node.right?.let(pending::add)
        }
        return shared
    }

    internal fun rootIdentityForTesting(): Any? = root

    internal fun nodeIdentityForTesting(key: K): Any? = findNode(key)

    internal fun cursorBoundRank(key: K, upper: Boolean): Int {
        var rank = 0
        var cursor = root
        while (cursor != null) {
            val comparison = keyComparator.compare(cursor.entry.key, key)
            if (comparison < 0 || (upper && comparison == 0)) {
                rank = Math.addExact(rank, Math.addExact(cursor.left?.count ?: 0, 1))
                cursor = cursor.right
            } else {
                cursor = cursor.left
            }
        }
        return rank
    }

    internal fun cursorEntryAt(rank: Int): PrioritySearchEntry<K, P, V>? {
        if (rank < 0 || rank >= count) return null
        var remaining = rank
        var cursor = root
        while (cursor != null) {
            val leftCount = cursor.left?.count ?: 0
            when {
                remaining < leftCount -> cursor = cursor.left
                remaining == leftCount -> return cursor.entry
                else -> {
                    remaining -= leftCount + 1
                    cursor = cursor.right
                }
            }
        }
        return null
    }

    private fun collectNodes(
        initialRoot: Node<K, P, V>?,
        destination: IdentityHashMap<Node<K, P, V>, Boolean>,
    ) {
        val pending = ArrayList<Node<K, P, V>>().apply { initialRoot?.let(::add) }
        while (pending.isNotEmpty()) {
            val node = pending.removeAt(pending.lastIndex)
            if (destination.put(node, true) != null) {
                continue
            }
            node.left?.let(pending::add)
            node.right?.let(pending::add)
        }
    }

    private fun findNode(key: K): Node<K, P, V>? {
        var node = root
        while (node != null) {
            val comparison = keyComparator.compare(key, node.entry.key)
            if (comparison == 0) {
                return node
            }
            node = if (comparison < 0) node.left else node.right
        }
        return null
    }

    private fun set(
        node: Node<K, P, V>?,
        entry: PrioritySearchEntry<K, P, V>,
        overwrite: Boolean,
    ): SetResult<K, P, V> {
        if (node == null) {
            return SetResult(newNode(entry, left = null, right = null), added = true, changed = true)
        }
        val comparison = keyComparator.compare(entry.key, node.entry.key)
        return when {
            comparison == 0 -> {
                if (!overwrite ||
                    (priorityComparator.compare(entry.priority, node.entry.priority) == 0 &&
                        psqValuesEqual(entry.priority, node.entry.priority) && psqValuesEqual(entry.value, node.entry.value))
                ) {
                    SetResult(node, added = false, changed = false)
                } else {
                    val retained = PrioritySearchEntry(node.entry.key, entry.priority, entry.value)
                    SetResult(newNode(retained, node.left, node.right), added = false, changed = true)
                }
            }
            comparison < 0 -> {
                val result = set(node.left, entry, overwrite)
                if (!result.changed) {
                    SetResult(node, result.added, changed = false)
                } else {
                    SetResult(
                        balance(newNode(node.entry, result.node, node.right)),
                        result.added,
                        changed = true,
                    )
                }
            }
            else -> {
                val result = set(node.right, entry, overwrite)
                if (!result.changed) {
                    SetResult(node, result.added, changed = false)
                } else {
                    SetResult(
                        balance(newNode(node.entry, node.left, result.node)),
                        result.added,
                        changed = true,
                    )
                }
            }
        }
    }

    private fun removeNode(node: Node<K, P, V>?, key: K): NodeRemoval<K, P, V> {
        node ?: return NodeRemoval(node = null, entry = null)
        val comparison = keyComparator.compare(key, node.entry.key)
        return when {
            comparison == 0 -> when {
                node.left == null -> NodeRemoval(node.right, node.entry)
                node.right == null -> NodeRemoval(node.left, node.entry)
                else -> {
                    val successor = minimumKeyNode(node.right)
                    val right = removeNode(node.right, successor.entry.key)
                    NodeRemoval(balance(newNode(successor.entry, node.left, right.node)), node.entry)
                }
            }
            comparison < 0 -> {
                val result = removeNode(node.left, key)
                if (result.entry == null) {
                    NodeRemoval(node, entry = null)
                } else {
                    NodeRemoval(balance(newNode(node.entry, result.node, node.right)), result.entry)
                }
            }
            else -> {
                val result = removeNode(node.right, key)
                if (result.entry == null) {
                    NodeRemoval(node, entry = null)
                } else {
                    NodeRemoval(balance(newNode(node.entry, node.left, result.node)), result.entry)
                }
            }
        }
    }

    private fun balance(node: Node<K, P, V>): Node<K, P, V> {
        val factor = heightOf(node.left) - heightOf(node.right)
        if (factor > 1) {
            val left = checkNotNull(node.left)
            return if (heightOf(left.left) < heightOf(left.right)) {
                rotateRight(newNode(node.entry, rotateLeft(left), node.right))
            } else {
                rotateRight(node)
            }
        }
        if (factor < -1) {
            val right = checkNotNull(node.right)
            return if (heightOf(right.right) < heightOf(right.left)) {
                rotateLeft(newNode(node.entry, node.left, rotateRight(right)))
            } else {
                rotateLeft(node)
            }
        }
        return node
    }

    private fun rotateLeft(node: Node<K, P, V>): Node<K, P, V> {
        val pivot = checkNotNull(node.right)
        return newNode(pivot.entry, newNode(node.entry, node.left, pivot.left), pivot.right)
    }

    private fun rotateRight(node: Node<K, P, V>): Node<K, P, V> {
        val pivot = checkNotNull(node.left)
        return newNode(pivot.entry, pivot.left, newNode(node.entry, pivot.right, node.right))
    }

    private fun newNode(
        entry: PrioritySearchEntry<K, P, V>,
        left: Node<K, P, V>?,
        right: Node<K, P, V>?,
    ): Node<K, P, V> = Node(entry, left, right, selectWinner(entry, left, right))

    private fun selectWinner(
        entry: PrioritySearchEntry<K, P, V>,
        left: Node<K, P, V>?,
        right: Node<K, P, V>?,
    ): PrioritySearchEntry<K, P, V> {
        var winner = entry
        if (left != null && isBefore(left.winner, winner)) {
            winner = left.winner
        }
        if (right != null && isBefore(right.winner, winner)) {
            winner = right.winner
        }
        return winner
    }

    private fun sameStoredEntry(
        left: PrioritySearchEntry<K, P, V>,
        right: PrioritySearchEntry<K, P, V>,
    ): Boolean =
        keyComparator.compare(left.key, right.key) == 0 && psqValuesEqual(left.key, right.key) &&
            priorityComparator.compare(left.priority, right.priority) == 0 && psqValuesEqual(left.priority, right.priority) &&
            psqValuesEqual(left.value, right.value)

    private fun isBefore(
        left: PrioritySearchEntry<K, P, V>,
        right: PrioritySearchEntry<K, P, V>,
    ): Boolean {
        val priority = priorityComparator.compare(left.priority, right.priority)
        return priority < 0 || (priority == 0 && keyComparator.compare(left.key, right.key) < 0)
    }

    private fun minimumKeyNode(initialNode: Node<K, P, V>): Node<K, P, V> {
        var node = initialNode
        while (true) {
            val left = node.left ?: return node
            node = left
        }
    }

    private fun heightOf(node: Node<K, P, V>?): Int = node?.height ?: 0

    private class Node<K, P, V>(
        val entry: PrioritySearchEntry<K, P, V>,
        val left: Node<K, P, V>?,
        val right: Node<K, P, V>?,
        val winner: PrioritySearchEntry<K, P, V>,
    ) {
        val height: Int = Math.addExact(1, max(left?.height ?: 0, right?.height ?: 0))
        val count: Int = Math.addExact(1, Math.addExact(left?.count ?: 0, right?.count ?: 0))
    }

    private data class SetResult<K, P, V>(
        val node: Node<K, P, V>,
        val added: Boolean,
        val changed: Boolean,
    )

    private data class NodeRemoval<K, P, V>(
        val node: Node<K, P, V>?,
        val entry: PrioritySearchEntry<K, P, V>?,
    )

    private data class QueryFrame<K, P, V>(val node: Node<K, P, V>, val emit: Boolean)

    private data class KeyBound<K>(val value: K)

    private data class ValidationFrame<K, P, V>(
        val node: Node<K, P, V>,
        val lower: KeyBound<K>?,
        val upper: KeyBound<K>?,
    )
}
